kratos/utilities/constraint.h

```diff
@@ -0,0 +1,294 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(CONSTRAINT_H)
+#define CONSTRAINT_H
+// System includes
+#include <vector>
+#include <unordered_map>
+#include <iostream>
+#include <tuple>
+#include <utility>
+#include <assert.h>
+
+// project includes
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include "geometries/geometry.h"
+#include "containers/constraint_data.h"
+#include "containers/variable.h"
+#include "containers/variable_component.h"
+#include "containers/vector_component_adaptor.h"
+
+namespace Kratos
+{
+/** \brief Constraint
+	* A class that implements the interface for different constraints to be applied on a system.
+    */
+
+class ProcessInfo;
+class Element;
+class Condition;
+
+template <class TSparseSpace,
+          class TDenseSpace //= DenseSpace<double>
+          >
+class Constraint

```

Comment:
This should class should not depend on other constraint classes.

Reply:
I agree. I do not see how is it depending on other constraint classes, can you point me to the lines ?

Reply:
It depends on ConstraintEquation and ConstraintEquationContainer because of the functions GetData() and GetConstraintEquation(). If you need these functions here then I'm fine with that, but it doesn't appear that the builder and solver needs to know about these classes. Therefore it introduces unnecessary code coupling when you include this in the builder and solver. You could consider placing these functions in another interface that doesn't need to get included in the builder and solver. That way you can still use them in multipoint constraint.

Reply:
Here I am a little bit confused about the design. If we have the builder and solver with constraint and constraint equation containers why we need this class also. It seems to me that its responsibility can be divided between two mentioned or you may merge the constraint equation container with this one.

Reply:
@pooyan-dadvand  the builder and solver need not know how the equations are to be modified by a constraint implementation/realization. So it should only use/call a set of functions (from the abstract class of constraint) to apply a constraint on the system. So there is a base class for all the constraint implementations/realizations. 
Coming to the ConstraintEquationContainer : A constraint is always defined as a set (can also be only one equation) of linear equations S = T*M + C . What to do with this equations is specific to implementation ... so I decided to divide them. As you say it is possible to merge the two (ConstraintEquationContainer and Constraint) then it will be a huge class with a "container" together with implementation specifics of how to use the contents (constraint equations) of that container. 

@msandre  In this process to make the container object (which is private) available to the child classes, I made the function GetData() in the base class. I see that we can put the GetConstraintEquation() in the multipoint_constraint class. This will not be a problem. Another approach is to derive the Constraint class from the ConstraintEquationContainer then the Constraint class will not have the redundant functions like GetData or GetConstraintEquation or AddConstraint (they will be inherited from the ConstraintEquationContainerClass).